### DISCUSSION TITLE
Add support for event-based shopping list syncing with other shopping list apps

### DISCUSSION THREAD

**Discussion by michael-genson** [mealie-recipes/mealie#1350] [Feature Request]
**Before submitting this feature request I have**
- [X] Reviewed the documentation and verified that my feature request isn't already a feature.
- [X] Reviewed the existing feature requests and my feature has not been requested.
- [X] Checked the [tasks taged](https://github.com/hay-kot/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

**Please Describe The Problem To Be Solved**
While the REST API opens up the possibility for accessing the shopping list and its items through querying, it would be ideal to implement a webhook or server-to-server notification for event-based shopping list synchronization with external apps (e.g. Alexa, Todoist).

**(Optional): Suggest A Solution**
A potentially simple solution would be to use the existing apprise integration service and extend support to shopping list items. There's already an event trigger for shopping list updates, but it doesn't capture adding or removing items. A quick-and-dirty solution would be to add the shopping list listener to item events, but that seems like a hack over having a separate event. One issue with the apprise integration is that it doesn't give any programmatic metadata (i.e. the id of the record that triggered the event), so either including the id of the record in the message, or including some metadata would be required.

Another option is to extend the webhook service to accommodate event-driven webhooks. This solves the issue of providing programmatic metadata, but would require some additional changes since they're currently scheduled, not event-based.

**Additional Information**
- [ ] If the feature is accepted I'm willing to submit a PR to provide this feature; I can probably figure out the backend, but the frontend is a bit of a mystery to me

**Comment by hay-kot** [comment]
We could just use the current Apprise integration and utilize the [Apprise JSON Type](https://github.com/caronc/apprise/wiki/Notify_Custom_JSON) and detect if it's using the JSON prefix and provide the data for each type of type of notification. I think this would allow for the most flexible approach and probably wouldn't require any Database changes, just some backend implementation - though they would be extensive ones changes to get the right data in the background.

**Reply by michael-genson** [reply]
Oh I see, so basically just customize the notification when using the JSON prefix. I can probably take a crack at that
